contracts/Launchpad.sol

```diff
@@ -0,0 +1,492 @@
+    }
+
+    modifier isTokenAddressSet(uint256 launchId) {
+        _isTokenAddressSet(launchId);
+        _;
+    }
+
+    function _isTokenAddressSet(uint256 launchId) internal view {
+        require(launchTokenAddress[launchId] != address(0), "Token address not set");
+    }
+
+    // VIEWS
+
+    function getTierForStakedAmount(uint256 amount) public view returns (uint256 tier) {
+        for (uint256 i = 1; i <= vars[VAR_TIERS_AMOUNT]; i++) {
+            if (amount >= tierStakingRequirement[i]) {
+                tier++;
+            } else {
+                break;
+            }
+        }
+    }
+
+    function getLaunchAllocationForTier(uint256 launchId, uint256 tier) public view returns (uint256) {
+        return launchBaseAllocation[launchId].mul(tierAllocationWeight[tier]);
+    }
+
+    function getUserRemainingAllocationForLaunch(address user, uint256 launchId) public view returns (uint256 maxAllocation) {
+        if(launchId % 2 == 0) {
+            maxAllocation = getLaunchAllocationForTier(launchId, getTierForStakedAmount(launchUserStakedAmountSnapshot[launchId.sub(1)][user]));
+        }
+        else {
+            maxAllocation = getLaunchAllocationForTier(launchId, getTierForStakedAmount(launchUserStakedAmountSnapshot[launchId][user]));
+        }
+        maxAllocation += launchUserUnclaimedSkillCommittedValue[launchId][user].mul(vars[VAR_UNCLAIMED_TO_ALLOCATION_MULTIPLIER]);
+        maxAllocation = maxAllocation.sub(launchUserInvestment[launchId][user]);
+    }
+
+    function getAvailableClaimAmount(uint256 launchId, uint256 vestingId) public onlyPhase1(launchId) view returns (uint256 claimAmount) {
+        uint256 decimals = ERC20(launchTokenAddress[launchId]).decimals();
+        uint256 totalUserInvestment = launchUserInvestment[launchId][msg.sender].add(launchUserInvestment[launchId + 1][msg.sender]);
+        claimAmount = totalUserInvestment.mul(1e18).div(launchTokenPrice[launchId]).mul(launchPeriodicVestingsPercentages[launchId][vestingId]).div(100);
+        claimAmount = Common.adjustDecimals(claimAmount, decimals);
+    }
+
+    function getLinearClaimAmount(uint256 launchId) public onlyPhase1(launchId) view returns (uint256 claimAmount) {
+        uint256 decimals = ERC20(launchTokenAddress[launchId]).decimals();
+        uint256 totalUserInvestment = launchUserInvestment[launchId][msg.sender].add(launchUserInvestment[launchId + 1][msg.sender]);
+        uint256 lastClaimTimestamp = userLinearVestingClaimTimestamp[msg.sender][launchId];
+        if(lastClaimTimestamp == 0) {
+            lastClaimTimestamp = launchLinearVestingsStartTimestamps[launchId];
+        }
+        claimAmount = totalUserInvestment.mul(1e18).div(launchTokenPrice[launchId]).mul(block.timestamp - lastClaimTimestamp).div(launchLinearVestingsDurations[launchId]);
+        claimAmount = Common.adjustDecimals(claimAmount, decimals);
+    }
+
+    function getLaunchBaseInfo(uint256 launchId) public view returns (
```

Comment:
I believe the launches mapping can be queried directly to return the struct as an array of values.
This can save contract space if needed, but there is no reason to change this otherwise (i think)

Example with quests struct:
frontend/src/store/quests/index.ts line 47
https://github.com/CryptoBlades/cryptoblades/blob/main/frontend/src/store/quests/index.ts
referencing SimpleQuests.sol struct mapping at line 94
https://github.com/CryptoBlades/cryptoblades/blob/main/contracts/SimpleQuests.sol